File: crates/pray-core/src/lockfile.rs
use crate::hashing::sha256_prefixed;
use crate::render::RenderedTarget;
use crate::{PrayError, PrayResult};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Component, Path, PathBuf};

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct Lockfile {
    pub prayfile_lock: String,
    pub spec: String,
    pub generated_by: String,
    pub manifest_hash: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub environment: Option<String>,
    pub source: Vec<LockSource>,
    pub package: Vec<LockedPackage>,
    pub target: Vec<LockedTarget>,
    pub managed_span: Vec<ManagedSpanRecord>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub provisioned: Vec<ProvisionedFileRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LockSource {
    pub name: String,
    pub kind: String,
    pub url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub revision: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host_key_fingerprint: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LockedPackage {
    pub name: String,
    pub version: String,
    pub source: Option<String>,
    pub path: String,
    pub tree_hash: String,
    pub artifact_hash: String,
    pub artifact: String,
    pub exports: Vec<String>,
    pub dependencies: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signer_fingerprint: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LockedTarget {
    pub name: String,
    pub outputs: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ManagedSpanRecord {
    pub id: String,
    pub target: String,
    pub open_line: usize,
    pub close_line: usize,
    pub ideal_checksum: String,
    pub package: String,
    pub export: String,
    pub source_checksum: String,
    pub silenced: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ProvisionedFileRecord {
    pub path: String,
    pub content_hash: String,
    pub package: String,
    pub export: String,
}

impl Lockfile {
    pub fn canonicalized(&self) -> Self {
        let mut lockfile = self.clone();
        lockfile
            .source
            .sort_by(|left, right| left.name.cmp(&right.name));
        lockfile.package.sort_by(|left, right| {
            left.name
                .cmp(&right.name)
                .then(left.source.cmp(&right.source))
                .then(left.version.cmp(&right.version))
        });
        lockfile
            .target
            .sort_by(|left, right| left.name.cmp(&right.name));
        lockfile.managed_span.sort_by(|left, right| {
            left.target
                .cmp(&right.target)
                .then(left.open_line.cmp(&right.open_line))
                .then(left.id.cmp(&right.id))
        });
        lockfile.provisioned.sort_by(|left, right| {
            left.path
                .cmp(&right.path)
                .then(left.package.cmp(&right.package))
        });
        lockfile
    }

    pub fn serialized(&self) -> PrayResult<String> {
        let bytes = toml::to_string_pretty(&self.canonicalized())
            .map_err(|error| PrayError::Manifest(error.to_string()))?;
        Ok(bytes)
    }

    pub fn file_hash(&self) -> PrayResult<String> {
        let text = self.serialized()?;
        Ok(sha256_prefixed(text.as_bytes()))
    }

    pub fn equivalent_to(&self, other: &Self) -> bool {
        self == &other.canonicalized()
    }
}

pub fn lockfiles_equivalent(canonical: &Lockfile, other: &Lockfile) -> bool {
    canonical.equivalent_to(other)
}

pub fn write_lockfile_if_changed(path: &Path, lockfile: &Lockfile) -> PrayResult<()> {
    let serialized = lockfile.serialized()?;
    if path.exists() {
        if let Ok(existing) = fs::read(path) {
            if existing == serialized.as_bytes() {
                return Ok(());
            }
        }
    }
    fs::write(path, serialized)?;
    Ok(())
}

pub fn write_lockfile(path: &Path, lockfile: &Lockfile) -> PrayResult<()> {
    let serialized = lockfile.serialized()?;
    fs::write(path, serialized)?;
    Ok(())
}

pub fn read_lockfile(path: &Path) -> PrayResult<Lockfile> {
    let text = fs::read_to_string(path)?;
    let lockfile = toml::from_str(&text).map_err(|error| PrayError::Parse {
        kind: "lockfile",
        message: error.to_string(),
    })?;
    Ok(lockfile)
}

pub fn relative_lockfile_path(project_root: &Path, path: &Path) -> String {
    let absolute = if path.is_absolute() {
        path.to_path_buf()
    } else {
        project_root.join(path)
    };
    let normalized_root = lexical_normalize_path(project_root);
    let normalized_absolute = lexical_normalize_path(&absolute);
    let relative = normalized_absolute
        .strip_prefix(&normalized_root)
        .map(Path::to_path_buf)
        .unwrap_or_else(|_| {
            if path.is_absolute() {
                path.to_path_buf()
            } else {
                lexical_normalize_path(path)
            }
        });
    format_relative_lockfile_path(&relative)
}

fn format_relative_lockfile_path(relative: &Path) -> String {
    let text = relative.to_string_lossy().replace('\\', "/");
    if text == "." || text.starts_with("./") {
        text
    } else {
        format!("./{text}")
    }
}

fn lexical_normalize_path(path: &Path) -> PathBuf {
    let mut normalized = PathBuf::new();
    for component in path.components() {
        match component {
            Component::Prefix(prefix) => normalized.push(prefix.as_os_str()),
            Component::RootDir => normalized.push(std::path::MAIN_SEPARATOR_STR),
            Component::CurDir => {}
            Component::ParentDir => {
                let _ = normalized.pop();
            }
            Component::Normal(segment) => normalized.push(segment),
        }
    }
    normalized
}

fn normalize_lockfile_artifact(project_root: &Path, artifact: &str, package_root: &Path) -> String {
    if let Some(path_text) = artifact.strip_prefix("path:") {
        let path = Path::new(path_text);
        let relative = if path.is_absolute() {
            relative_lockfile_path(project_root, path)
        } else {
            relative_lockfile_path(project_root, package_root)
        };
        format!("path:{relative}")
    } else {
        artifact.to_string()
    }
}

#[allow(clippy::too_many_arguments)]
pub fn build_lockfile(
    manifest_hash: String,
    environment: Option<String>,
    project_root: &Path,
    manifest_sources: &[crate::manifest::ManifestSource],
    manifest_targets: &[crate::manifest::ManifestTarget],
    rendered: &[RenderedTarget],
    packages: &[crate::resolve::ResolvedPackage],
    source_revisions: &std::collections::BTreeMap<String, String>,
    source_host_keys: &std::collections::BTreeMap<String, String>,
) -> Lockfile {
    Lockfile {
        prayfile_lock: "1".to_string(),
        spec: "0.1".to_string(),
        generated_by: format!("pray {}", env!("CARGO_PKG_VERSION")),
        manifest_hash,
        environment,
        source: manifest_sources
            .iter()
            .map(|source| LockSource {
                name: source.name.clone(),
                kind: source.kind.clone(),
                url: source.url.clone(),
                revision: source_revisions.get(&source.name).cloned(),
                host_key_fingerprint: source_host_keys.get(&source.name).cloned(),
            })
            .collect(),
        package: packages
            .iter()
            .map(|package| LockedPackage {
                name: package.declaration.name.clone(),
                version: package.spec.version.clone(),
                source: package.declaration.source.clone(),
                path: relative_lockfile_path(project_root, &package.root),
                tree_hash: package.tree_hash.clone(),
                artifact_hash: package.artifact_hash.clone(),
                artifact: normalize_lockfile_artifact(
                    project_root,
                    &package.artifact,
                    &package.root,
                ),
                exports: package.selected_exports.clone(),
                dependencies: package
                    .spec
                    .dependencies
                    .iter()
                    .map(|dependency| dependency.name.clone())
                    .collect(),
                signer_fingerprint: package.signer_fingerprint.clone(),
            })
            .collect(),
        target: manifest_targets
            .iter()
            .map(|target| LockedTarget {
                name: target.name.clone(),
                outputs: target.outputs.clone(),
            })
            .collect(),
        managed_span: rendered
            .iter()
            .flat_map(|target| target.managed_spans.iter().cloned())
            .collect(),
        provisioned: Vec::new(),
    }
    .canonicalized()
}

#[cfg(test)]
#[path = "lockfile_unit.rs"]
mod lockfile_unit;

File: rubygems/pray-cli/lib/pray/lockfile.rb
# frozen_string_literal: true

require "toml-rb"

require_relative "lockfile_serialize"

module Pray
  LockSource = Struct.new(:name, :kind, :url, :revision, :host_key_fingerprint, keyword_init: true)
  LockedPackage = Struct.new(
    :name, :version, :source, :path, :tree_hash, :artifact_hash, :artifact,
    :exports, :dependencies, :signer_fingerprint,
    keyword_init: true
  ) do
    def initialize(exports: [], dependencies: [], signer_fingerprint: nil, source: nil, **kwargs)
      super(**kwargs, exports: exports, dependencies: dependencies, signer_fingerprint: signer_fingerprint, source: source)
    end
  end

  LockedTarget = Struct.new(:name, :outputs, keyword_init: true)
  ManagedSpanRecord = Struct.new(
    :id, :target, :open_line, :close_line, :ideal_checksum, :package, :export,
    :source_checksum, :silenced,
    keyword_init: true
  )

  Lockfile = Struct.new(
    :prayfile_lock, :spec, :generated_by, :manifest_hash, :environment, :source, :package,
    :target, :managed_span,
    keyword_init: true
  ) do
    def initialize(
      prayfile_lock: "1", spec: "0.1", generated_by: Pray::GENERATED_BY,
      manifest_hash: "", environment: nil, source: [], package: [], target: [], managed_span: []
    )
      super
    end

    def canonicalized
      dup.tap do |copy|
        copy.source = source.sort_by(&:name)
        copy.package = package.sort_by { |entry| [entry.name, entry.source.to_s, entry.version] }
        copy.target = target.sort_by(&:name)
        copy.managed_span = managed_span.sort_by { |span| [span.target, span.open_line, span.id] }
      end
    end

    def serialized
      LockfileIO.lockfile_to_toml(canonicalized)
    end

    def file_hash
      Hashing.sha256_prefixed(serialized)
    end

    def equivalent_to?(other)
      canonicalized == other.canonicalized
    end
  end

  module LockfileIO
    module_function

    def lockfile_to_toml(lockfile)
      LockfileSerialize.lockfile_to_toml(lockfile)
    end

    def lockfile_to_hash(lockfile)
      {
        "prayfile_lock" => lockfile.prayfile_lock,
        "spec" => lockfile.spec,
        "generated_by" => lockfile.generated_by,
        "manifest_hash" => lockfile.manifest_hash,
        "source" => lockfile.source.map { |entry| source_to_hash(entry) },
        "package" => lockfile.package.map { |entry| package_to_hash(entry) },
        "target" => lockfile.target.map { |entry| { "name" => entry.name, "outputs" => entry.outputs } },
        "managed_span" => lockfile.managed_span.map { |entry| managed_span_to_hash(entry) }
      }
    end

    def source_to_hash(entry)
      hash = { "name" => entry.name, "kind" => entry.kind, "url" => entry.url }
      hash["revision"] = entry.revision if entry.revision
      hash["host_key_fingerprint"] = entry.host_key_fingerprint if entry.host_key_fingerprint
      hash
    end

    def package_to_hash(entry)
      hash = {
        "name" => entry.name,
        "version" => entry.version,
        "path" => entry.path,
        "tree_hash" => entry.tree_hash,
        "artifact_hash" => entry.artifact_hash,
        "artifact" => entry.artifact,
        "exports" => entry.exports,
        "dependencies" => entry.dependencies
      }
      hash["source"] = entry.source unless entry.source.nil?
      hash["signer_fingerprint"] = entry.signer_fingerprint if entry.signer_fingerprint
      hash
    end

    def managed_span_to_hash(entry)
      {
        "id" => entry.id,
        "target" => entry.target,
        "open_line" => entry.open_line,
        "close_line" => entry.close_line,
        "ideal_checksum" => entry.ideal_checksum,
        "package" => entry.package,
        "export" => entry.export,
        "source_checksum" => entry.source_checksum,
        "silenced" => entry.silenced
      }
    end

    def parse_lockfile(text)
      data = TomlRB.parse(text)
      from_hash(data)
    rescue TomlRB::ParseError => error
      raise Error.parse("lockfile", error.message)
    end

    def read_lockfile(path)
      parse_lockfile(File.read(path))
    end

    def serialize_lockfile(lockfile)
      lockfile_to_toml(lockfile.canonicalized)
    end

    def lockfile_hash(lockfile)
      Hashing.sha256_prefixed(serialize_lockfile(lockfile))
    end

    def write_lockfile(path, lockfile)
      File.write(path, serialize_lockfile(lockfile))
    end

    def write_lockfile_if_changed(path, lockfile)
      serialized = serialize_lockfile(lockfile)
      if File.exist?(path) && File.binread(path) == serialized
        return
      end

      File.write(path, serialized)
    end

    def lockfiles_equivalent?(left, right)
      left.equivalent_to?(right)
    end

    def build_lockfile(manifest_hash, environment, project_root, manifest_sources, manifest_targets, rendered, packages, source_revisions, source_host_keys)
      Lockfile.new(
        manifest_hash: manifest_hash,
        environment: environment,
        source: manifest_sources.map do |source|
          LockSource.new(
            name: source.name,
            kind: source.kind,
            url: source.url,
            revision: source_revisions[source.name],
            host_key_fingerprint: source_host_keys[source.name]
          )
        end,
        package: packages.map do |package|
          LockedPackage.new(
            name: package.declaration.name,
            version: package.spec.version,
            source: package.declaration.source,
            path: relative_lockfile_path(project_root, package.root),
            tree_hash: package.tree_hash,
            artifact_hash: package.artifact_hash,
            artifact: normalize_lockfile_artifact(project_root, package.artifact, package.root),
            exports: package.selected_exports,
            dependencies: package.spec.dependencies.map(&:name),
            signer_fingerprint: package.signer_fingerprint
          )
        end,
        target: manifest_targets.map { |target| LockedTarget.new(name: target.name, outputs: target.outputs) },
        managed_span: rendered.flat_map(&:managed_spans)
      ).canonicalized
    end

    def from_hash(data)
      Lockfile.new(
        prayfile_lock: data["prayfile_lock"],
        spec: data["spec"],
        generated_by: data["generated_by"],
        manifest_hash: data["manifest_hash"],
        environment: data["environment"],
        source: Array(data["source"]).map do |entry|
          LockSource.new(
            name: entry["name"],
            kind: entry["kind"],
            url: entry["url"],
            revision: entry["revision"],
            host_key_fingerprint: entry["host_key_fingerprint"]
          )
        end,
        package: Array(data["package"]).map do |entry|
          LockedPackage.new(
            name: entry["name"],
            version: entry["version"],
            source: entry["source"],
            path: entry["path"],
            tree_hash: entry["tree_hash"],
            artifact_hash: entry["artifact_hash"],
            artifact: entry["artifact"],
            exports: entry["exports"] || [],
            dependencies: entry["dependencies"] || [],
            signer_fingerprint: entry["signer_fingerprint"]
          )
        end,
        target: Array(data["target"]).map { |entry| LockedTarget.new(name: entry["name"], outputs: entry["outputs"] || []) },
        managed_span: Array(data["managed_span"]).map do |entry|
          ManagedSpanRecord.new(
            id: entry["id"],
            target: entry["target"],
            open_line: entry["open_line"],
            close_line: entry["close_line"],
            ideal_checksum: entry["ideal_checksum"],
            package: entry["package"],
            export: entry["export"],
            source_checksum: entry["source_checksum"],
            silenced: entry["silenced"]
          )
        end
      )
    end

    def relative_lockfile_path(project_root, path)
      absolute = Pathname(path).absolute? ? Pathname(path) : Pathname(project_root).join(path)
      normalized_root = Pathname(project_root).cleanpath
      normalized_absolute = absolute.cleanpath
      relative = normalized_absolute.relative_path_from(normalized_root)
      format_relative_lockfile_path(relative)
    rescue ArgumentError
      format_relative_lockfile_path(Pathname(path).cleanpath)
    end

    def format_relative_lockfile_path(relative)
      text = relative.to_s.tr("\\", "/")
      text == "." || text.start_with?("./") ? text : "./#{text}"
    end

    def normalize_lockfile_artifact(project_root, artifact, package_root)
      return artifact unless artifact.start_with?("path:")

      path_text = artifact.delete_prefix("path:")
      path = Pathname(path_text)
      relative = if path.absolute?
                   relative_lockfile_path(project_root, path)
                 else
                   relative_lockfile_path(project_root, package_root)
                 end
      "path:#{relative}"
    end
  end

  extend LockfileIO
end
